scripts/create_services_for_territories.rb

```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+
+# Ce script a pour vocation de créer les services en spécifiant leur territoire, sur la base des services existant
+# Pour chaque service, identifier les territoires liés
+# Pour chaque service:
+#  - Pour chaque territoire du service:
+#     - Dupliquer le service dans le territoire
+#     - Mettre à jour tous les agents du territoire avec le nouveau service
+#     - Mettre à jour tous les motifs du territoire avec le nouveau service
+#
+# Supprimer tous les services sans TerritoryID
+
+services = Service.joins(
+  agents: { organisations: :territory }
+).select(
+  :id, :name, :short_name, "ARRAY_AGG(DISTINCT territories.id) as territories_ids"
+).group(
+  :id, :name, :short_name
+)
+
+# rubocop:disable Rails/SkipsModelValidations
+Service.transaction do
+  services.each do |service|
+    service.territories_ids.each do |territory_id|
+      new_service = service.dup
+      new_service.territory_id = territory_id
+
+      # This is to avoid issues with the unique validation on name and short_name columns
+      service.update!(name: "old_#{service.name}", short_name: "old_#{service.short_name}")
+      new_service.save!
+
+      Agent.joins(organisations: :territory).where(service_id: service.id, organisations: { territory_id: territory_id }).update_all(service_id: new_service.id)
+      Motif.joins(organisation: :territory).where(service_id: service.id, organisation: { territory_id: terr }).update_all(service_id: new_service.id)
+    end
+  end
+
+  Service.where(territory_id: nil).destroy_all
```

Comment:
En lançant le script sur la base de données de seeds, cette ligne supprime de nombreux `Motif` et `MotifsPlageOuverture` en cascade, car on a `has_many :motifs, dependent: :destroy` dans `service.rb`. 

J'ai donc l'impression qu'on a des motifs qui restent non migrés en fin de script. Ça s'expliquer par le fait qu'on itère sur les services et les territoires à travers une jointure sur les agents, et il doit y avoir des territoires qui contiennent des motifs lié à un service dont aucun agent du territoire ne fait partie.

D'instinct je me dis que de faire une double boucle par service puis par territoire, ça se joue. Et pour chaque combinaison se service et territoire, on crée un nouveau service si c'est nécessaire. C'est une approche plus lente mais plus simple et lisible que la belle jointure au début de ce fichier. :wink:

Reply:
J'ai regardé : seulement 472 agents sur 2322 ont des RDVs. Une possibilité est de les supprimer, mais ça fait quand-même 58059 RDVs ! :scream: 

On discute demain. :call_me_hand: